procrastinate/psycopg2_connector.py

```diff
@@ -34,6 +34,45 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def wrap_query_exceptions(func: Callable) -> Callable:
+    """
+    Detect psycopg2 InterfaceError's with a "connection already closed" message and
+    retry a number of times.
+
+    This is to handle the case where the database connection (obtained from the pool)
+    was actually closed by the server. In this case, pyscopg2 raises an InterfaceError
+    with a "connection already closed" message (and no pgcode) when the connection is
+    used for issuing a query. What we do is retry when an InterfaceError is raised, and
+    until max_tries is reached.
+
+    max_tries is set to the max pool size + 1 (maxconn + 1) to handle the case where
+    all the connections we have in the pool were closed by the server.
+    """
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        final_exc = None
+        max_tries = (
+            args[0]._pool.maxconn + 1 if args and getattr(args[0], "_pool", None) else 1
+        )
+        for _ in range(max_tries):
+            try:
+                return func(*args, **kwargs)
+            except psycopg2.errors.InterfaceError as exc:
```

Comment:
I thought we had a AdminShutdown exception ?

Reply:
This is what we get:

```
Traceback (most recent call last):
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 168, in _connection
    yield connection
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 183, in execute_query_one
    cursor.execute(query, self._wrap_json(arguments))
  File "/home/elemoine/.virtualenvs/procrastinate/lib/python3.8/site-packages/psycopg2/extras.py", line 248, in execute
    return super(RealDictCursor, self).execute(query, vars)
psycopg2.errors.AdminShutdown: terminating connection due to administrator command
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 25, in wrapped
    return func(*args, **kwargs)
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 184, in execute_query_one
    return cursor.fetchone()
  File "/usr/lib64/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 168, in _connection
    yield connection
psycopg2.InterfaceError: connection already closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/elemoine/src/procrastinate/procrastinate/tasks.py", line 112, in defer
    return self.configure().defer(**task_kwargs)
  File "/home/elemoine/src/procrastinate/procrastinate/jobs.py", line 151, in defer
    id = self.job_store.defer_job(job=job)
  File "/home/elemoine/src/procrastinate/procrastinate/store.py", line 35, in defer_job
    result = self.connector.execute_query_one(
  File "/home/elemoine/src/procrastinate/procrastinate/psycopg2_connector.py", line 29, in wrapped
    raise exceptions.ConnectorException from exc
procrastinate.exceptions.ConnectorException: 
    Database error.
```

Reply:
Ok, I think this hides something. Back to the black board…

Reply:
Now handling AdminShutdown exceptions rather than InterfaceError. This made me rewrite the `_connection` context manager function to avoid rolling back the connection or putting it back to the pool when an AdminShutdown error occurs. Rolling back the connection or putting it back to the pool is what makes psycopg2 raise an InterfaceError.

This was more complex than I initially expected.